neat_core.c

```diff
@@ -804,6 +1035,210 @@ neat_open(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port)
+}
+
+neat_error_code
+neat_open_multistream(neat_ctx *mgr, neat_flow *flow, const char *name, uint16_t port, const char* localname, int count)
+{
+    neat_log(NEAT_LOG_DEBUG, "%s", __func__);
+
+    if (count < 1) {
+        return NEAT_ERROR_BAD_ARGUMENT;
+    } else if (count > 1) {
+	// TODO: multi-streaming is not yet available from the SCTP
+	// layer.
+	return NEAT_ERROR_UNABLE;
+    }
+
+    return neat_open_localname(mgr, flow, name, port, localname);
+}
+
+neat_error_code
+neat_change_timeout(neat_ctx *mgr, neat_flow *flow, int seconds)
+{
+    neat_log(NEAT_LOG_DEBUG, "%s", __func__);
+
+    // TCP User Timeout isn't supported by these platforms:
+#if !(defined(__FreeBSD__) || defined(__NetBSD__) || defined (__APPLE__))
+    if (seconds < 0) {
+            neat_log(NEAT_LOG_WARNING,
+                    "Unable to change timeout: "
+                    "Negative timeout specified");
+
+            return NEAT_ERROR_BAD_ARGUMENT;
+    }
+
+    if (flow->sockProtocol == IPPROTO_TCP) {
+        if (flow->fd == -1) {
+            neat_log(NEAT_LOG_WARNING,
+                    "Unable to change timeout for TCP socket: "
+                    "Invalid socket value");
+            return NEAT_ERROR_BAD_ARGUMENT;
+        }
+
+        unsigned int new_value =
+#if defined(__FreeBSD__) || defined(__NetBSD__)
+            ((unsigned int)seconds);
+#else
+            ((unsigned int)seconds) * 1000; // seconds -> ms
+#endif
+
+        int rc = setsockopt(flow->fd, IPPROTO_TCP,
+#if defined(__FreeBSD__) || defined(__NetBSD__)
+	    // KAH: turns out this was implemented on a feature
+	    // branch, and sadly not merged to mainline
+                TCP_SNDUTO_TIMEOUT,
+#else
+                TCP_USER_TIMEOUT,
+#endif
+                &new_value,
+                sizeof(new_value));
+
+        if (rc < 0) {
+            neat_log(NEAT_LOG_ERROR,
+                    "Unable to change timeout for TCP socket: "
+                    "Call to setsockopt failed with errno=%d", errno);
+            return NEAT_ERROR_IO;
+        }
+
+        return NEAT_ERROR_OK;
+    } else if (flow->sockProtocol == IPPROTO_SCTP) {
+#ifdef USRSCTP_SUPPORT
+        unsigned int new_value = ((unsigned int)seconds) * 1000 / 4; // seconds -> ms
+        usrsctp_sysctl_set_sctp_heartbeat_interval_default((uint32_t)new_value);
+        usrsctp_sysctl_set_sctp_assoc_rtx_max_default(4);
+#else
+        struct sctp_paddrparams params;
+        unsigned int optsize = sizeof(params);
+        int rc = getsockopt(flow->fd, IPPROTO_SCTP, SCTP_PEER_ADDR_PARAMS, &params, &optsize);
+        if (rc < 0) {
+            neat_log(NEAT_LOG_ERROR,
+                    "Unable to change timeout for TCP socket: "
+                    "Call to getsockopt failed with errno=%d", errno);
+            return NEAT_ERROR_IO;
+        }
+
+        params.spp_hbinterval = (unsigned int)seconds * 1000 / 4;
+        params.spp_flags |= SPP_HB_ENABLE;
+        params.spp_pathmaxrxt = 4;
+
```

Comment:
I don't understand these settings. This looks like magic behaviour to the user. I think this is a TCP only feature available only on one platform. So fail this call on other protocols or platforms.